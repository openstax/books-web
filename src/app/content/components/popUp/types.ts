import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
