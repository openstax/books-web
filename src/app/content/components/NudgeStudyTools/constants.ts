export const arrowDesktopHeight = 13;
export const arrowMobileHeight = 6.2;
export const arrowLeftMargin = 1;
export const arrowTopMargin = 1;

export const closeButtonDistanceFromContent = 4;
export const closeButtonMobileMargin = 3;

export const contentMarginTop = 1;

export const spotlightPadding = 1;

export const cookieNudge = {
  // How many times nuge was shown
  counter: 'nudge_study_guides_counter',
  // When nudge was shown
  date: 'nudge_study_guides_date',
  // How many pages user has opened (opening the same page twice counts it as 2)
  pageCounter: 'nudge_study_guides_page_counter',
};

// Show nudge only for users which opened at least 2 pages
export const nudgeStudyToolsMinPageLimit = 2;

// Show nudge every 4 weeks: 4 * 7 days * 24 hours * 60 minutes * 60 seconds * 1000 ms
export const timeIntervalBetweenShowingNudgeInMs = 4 * 7 * 24 * 60 * 60 * 1000;
// Show nudge max 3 times
export const nudgeStudyToolsShowLimit = 3;
// Id of the element which should be highlighted
export const nudgeStudyToolsTargetId = 'nudge-study-tools';
