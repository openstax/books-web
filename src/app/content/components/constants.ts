import theme from '../../theme';

export const bookBannerDesktopBigHeight = 13;
export const bookBannerDesktopMiniHeight = 7;
export const bookBannerMobileBigHeight = 10.4;
export const bookBannerMobileMiniHeight = 6;

export const sidebarDesktopWidth = 33.5;
export const sidebarMobileWidth = 28.8;
export const sidebarTransitionTime = 300;

export const toolbarIconColor = theme.color.primary.gray;
export const toolbarDesktopHeight = 5;
export const toolbarMobileHeight = 4;

export const contentTextWidth = 84;

export const mainContentBackground = '#fff';

export const maxContentGutter = 6;
export const contentWrapperMaxWidth = contentTextWidth + sidebarDesktopWidth + maxContentGutter * 2;
