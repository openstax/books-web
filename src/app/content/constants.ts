import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import theme from '../theme';

export const studyGuidesFeatureFlag = 'studyGuidesEnabled';
export const practiceQuestionsFeatureFlag = 'practiceQuestionsEnabled';

export const maxHighlightsApiPageSize = 200;
export const maxResourcesPerFetch = 10;

export const summaryPageSize = 20;
export const loadMoreDistanceFromBottom = 50;

export const modalQueryParameterName = 'modal';

export const highlightStyles: Array<{label: HighlightColorEnum, passive: string, focused: string}> = [
  {label: HighlightColorEnum.Yellow, passive: '#ffff8a', focused: '#fed200'},
  {label: HighlightColorEnum.Green, passive: '#def99f', focused: '#92d101'},
  {label: HighlightColorEnum.Blue, passive: '#c8f5ff', focused: '#00c3ed'},
  {label: HighlightColorEnum.Purple, passive: '#cbcfff', focused: '#545ec8'},
  {label: HighlightColorEnum.Pink, passive: '#ffc5e1', focused: '#de017e'},
];

const getPrimaryBaseColors = (primaryColors: {[key: string]: {[key: string]: string}}) => {
  const primaryBases: { [key: string]: string } = {};

  // tslint:disable-next-line: max-line-length
  Object.keys(primaryColors).forEach((color) => {primaryBases[color] = primaryColors[color].base; });
  return primaryBases;
};

export const searchButtonStyles: {[key: string]: string} = {
  gray: '#5D6062',
  transparent: 'transparent',
  ...getPrimaryBaseColors(theme.color.primary) as {},
};
