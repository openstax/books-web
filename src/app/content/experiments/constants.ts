export const experiments: { [key: string]: any } = {
  searchButton: [
    'originalButtonEnabled',
    'bannerColorButtonEnabled',
    'grayButtonEnabled',
  ],
};

export const experimentIds: { [key: string]: any } = {
  OCCkMMCZSwW87szzpniCow: 'searchButton',
};
