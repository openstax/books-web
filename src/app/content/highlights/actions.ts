import { createStandardAction } from 'typesafe-actions';

export const enableHighlighting = createStandardAction('FeatureFlag/enableHighlighting')<void>();
