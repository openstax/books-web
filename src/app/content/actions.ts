import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as searchActions from './search/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();

export const closeCallToActionPopup = createStandardAction('Content/closeCallToActionPopup')();
export const openCallToActionPopup = createStandardAction('Content/openCallToActionPopup')();
