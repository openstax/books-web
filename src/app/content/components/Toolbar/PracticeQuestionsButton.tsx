import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import practiceQuestionsIcon from '../../../../assets/practiceQuestionsIcon.svg';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import { openPracticeQuestions } from '../../practiceQuestions/actions';
import { practiceQuestionsEnabled } from '../../practiceQuestions/selectors';
import { PlainButton, toolbarDefaultButton, toolbarDefaultText } from './styled';

// tslint:disable-next-line:variable-name
export const PracticeQuestionsWrapper = styled(PlainButton)`
  ${toolbarDefaultButton}
`;

// tslint:disable-next-line:variable-name
const PracticeQuestionsIcon = styled.img`
  padding: 0.2rem;
`;

// tslint:disable-next-line:variable-name
const PracticeQuestionsText = styled.span`
  ${toolbarDefaultText}
`;

// tslint:disable-next-line:variable-name
const PracticeQuestionsButton = () => {
  const dispatch = useDispatch();
  const isEnabled = useSelector(practiceQuestionsEnabled);
  const trackOpenClose = useAnalyticsEvent('openClosePracticeQuestions');

  if (!isEnabled) { return null; }

  const openPracticeQuestionsSummary = () => {
    dispatch(openPracticeQuestions());
    trackOpenClose();
  };

  return <FormattedMessage id='i18n:toolbar:practice-questions:button:text'>
    {(msg: Element | string) =>
      <PracticeQuestionsWrapper onClick={openPracticeQuestionsSummary} aria-label={msg}>
        <PracticeQuestionsIcon aria-hidden='true' src={practiceQuestionsIcon} />
        <PracticeQuestionsText>{msg}</PracticeQuestionsText>
      </PracticeQuestionsWrapper>
    }
  </FormattedMessage>;
};

export default PracticeQuestionsButton;
