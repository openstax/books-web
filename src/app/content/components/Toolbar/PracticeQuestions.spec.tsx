import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { receiveFeatureFlags } from '../../../actions';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { practiceQuestionsFeatureFlag } from '../../constants';
import PracticeQuestionsButton, { PracticeQuestionsWrapper } from './PracticeQuestionsButton';

describe('practice questions button', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;
  let render: () => JSX.Element;

  beforeEach(() => {
    store = createTestStore();
    services = createTestServices();
    render = () => <Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <PracticeQuestionsButton />
        </MessageProvider>
      </Services.Provider>
    </Provider>;
  });

  it('does not render if feature flag is not enabled', () => {
    store.dispatch(receiveFeatureFlags([]));

    const component = renderer.create(render());

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('render if feature flag is enabled', () => {
    store.dispatch(receiveFeatureFlags([practiceQuestionsFeatureFlag]));

    const component = renderer.create(render());

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('clicking button works', () => {
    store.dispatch(receiveFeatureFlags([practiceQuestionsFeatureFlag]));

    const component = renderer.create(render());

    renderer.act(() => {
      const button = component.root.findByType(PracticeQuestionsWrapper);
      button.props.onClick();
    });

    // TODO: Add here something to expect in another PR
  });
});
