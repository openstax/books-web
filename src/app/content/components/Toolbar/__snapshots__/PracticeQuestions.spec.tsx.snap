// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`practice questions button does not render if feature flag is not enabled and there are no practice questions 1`] = `null`;

exports[`practice questions button renders if feature flag is enabled and there are practice questions 1`] = `null`;
